public/app/plugins/datasource/elasticsearch/components/hooks/useCustomValue.ts

```diff
@@ -0,0 +1,47 @@
+import { SelectableValue } from '@grafana/data';
+import { Select } from '@grafana/ui';
+import { ComponentProps, useState } from 'react';
+
+const hasValue = <T extends SelectableValue>(searchValue: T['value']) => ({ value }: T) => value === searchValue;
+
+const getInitialState = <T extends SelectableValue>(initialOptions: T[], initialValue?: T['value']): T[] => {
+  return initialOptions.concat(
+    initialOptions.some(hasValue(initialValue))
+      ? []
+      : ({
+          value: initialValue,
+          label: initialValue,
+        } as T)
+  );
+};
+
+interface UseCustomValue<T extends SelectableValue> {
+  options: T[];
+  value?: T['value'];
+  onChange?: (value: T['value']) => void;
+}
+export const useCustomValue = <T extends SelectableValue>({
```

Comment:
I'm having a hard time seeing the value adding this hook. Which logic feels duplicated to you?

Reply:
basically, when you create a custom value in a select, you also need to persist it somewhere, ie. by merging a set of default options and the custom created ones. then, when first mounting the component, the `options` prop should already have an option for what you set in `value`, otherwise the select will look like in an empty state.
also, when creating a new option you manually need to call the onchange because the select won't do it for you, i can see this being useful in some situations, but in the ES query editor this is not the case.

basically you'd have to implement what i did in https://github.com/grafana/grafana/pull/36168/files#diff-41c4e95630395cc02478c268278f324d4f8d61d3eeec23fbb857fa1402df4b9cR45-R63 for every select that allows custom values, and instead of testing it each time, this would basically allow me to implement the logic once and reuse it.

It is only implemented for `date_histogram` and `terms` aggregation editors now, but this should be done for all of the fields that allow custom values.

Reply:
to add a bit more context, the other approach would have been to create a wrapper component that accepts the same props as the `Select` component in `@grafana/ui` and adds this logic, the reason i went for the hook approach is that i feel it's better suited for this case because i don't need to change the component behaviour, i just need to apply the same logic to it.

Reply:
Alright, I think what you said makes sense.